README
$Id: README,v 1.1.1.1 2004/08/15 14:55:43 btrott Exp $

This is Class::ErrorHandler, a base class for classes that need to do
error handling (which is, probably, most of them).

PREREQUISITES

None.

INSTALLATION

Class::ErrorHandler installation is straightforward. If your CPAN shell
is set up, you should just be able to do

    % perl -MCPAN -e 'install Class::ErrorHandler'

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

Benjamin Trott / cpan@stupidfool.org


Changes
# $Id: Changes,v 1.1.1.1 2004/08/15 14:55:43 btrott Exp $

Revision history for Class::ErrorHandler

{{$NEXT}}

0.04 2015-05-10T22:08:23Z

    - use EU::MM

0.03 2013-08-14T03:15:44Z

    - minl migrate

0.02  2013.04.25
    - Re-packaging without auto_install().
      (ANDK)

0.01  2004.08.15
    - Initial distribution.


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Class-ErrorHandler',
    module_name     => 'Class::ErrorHandler',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


Makefile.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;
use strict;

use ExtUtils::MakeMaker;


my %WriteMakefileArgs = (
    NAME     => 'Class::ErrorHandler',
    DISTNAME => 'Class-ErrorHandler',
    VERSION  => '0.03',
    CONFIGURE_REQUIRES => {
  "ExtUtils::MakeMaker" => 0
}
,
    BUILD_REQUIRES     => {}
,
    TEST_REQUIRES      => {}
,
    PREREQ_PM          => {
  "perl" => "5.008_001"
}
,
);

my $full_prereqs = {
  "ExtUtils::MakeMaker" => 0,
  "perl" => "5.008_001"
}
;

unless (eval { ExtUtils::MakeMaker->VERSION(6.63_03) }) {
    delete $WriteMakefileArgs{TEST_REQUIRES};
    delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM} = $full_prereqs;
}

unless (eval { ExtUtils::MakeMaker->VERSION(6.52) }) {
    delete $WriteMakefileArgs{CONFIGURE_REQUIRES};
}

unless (eval { ExtUtils::MakeMaker->VERSION(6.57_01) }) {
    use File::Copy;
    copy('META.yml', 'MYMETA.yml')   or warn "META.yml: $!";
    copy('META.json', 'MYMETA.json') or warn "META.json: $!";
}

WriteMakefile(%WriteMakefileArgs);


lib/Class/ErrorHandler.pm
# $Id: ErrorHandler.pm,v 1.1.1.1 2004/08/15 14:55:43 btrott Exp $

package Class::ErrorHandler;
use strict;

use vars qw( $VERSION $ERROR );
$VERSION = '0.02';

sub error {
    my $msg = $_[1] || '';
    if (ref($_[0])) {
        $_[0]->{_errstr} = $msg;
    } else {
        $ERROR = $msg;
    }
    return;
}

sub errstr {
    ref($_[0]) ? $_[0]->{_errstr} : $ERROR
}

1;
__END__

=head1 NAME

Class::ErrorHandler - Base class for error handling

=head1 SYNOPSIS

    package Foo;
    use base qw( Class::ErrorHandler );

    sub class_method {
        my $class = shift;
        ...
        return $class->error("Help!")
            unless $continue;
    }

    sub object_method {
        my $obj = shift;
        ...
        return $obj->error("I am no more")
            unless $continue;
    }

    package main;
    use Foo;

    Foo->class_method or die Foo->errstr;

    my $foo = Foo->new;
    $foo->object_method or die $foo->errstr;

=head1 DESCRIPTION

I<Class::ErrorHandler> provides an error-handling mechanism that's generic
enough to be used as the base class for a variety of OO classes. Subclasses
inherit its two error-handling methods, I<error> and I<errstr>, to
communicate error messages back to the calling program.

On failure (for whatever reason), a subclass should call I<error> and return
to the caller; I<error> itself sets the error message internally, then
returns C<undef>. This has the effect of the method that failed returning
C<undef> to the caller. The caller should check for errors by checking for a
return value of C<undef>, and calling I<errstr> to get the value of the
error message on an error.

As demonstrated in the L<SYNOPSIS>, I<error> and I<errstr> work as both class
methods and object methods.

=head1 USAGE

=head2 Class->error($message)

=head2 $object->error($message)

Sets the error message for either the class I<Class> or the object
I<$object> to the message I<$message>. Returns C<undef>.

=head2 Class->errstr

=head2 $object->errstr

Accesses the last error message set in the class I<Class> or the
object I<$object>, respectively, and returns that error message.

=head1 LICENSE

I<Class::ErrorHandler> is free software; you may redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR & COPYRIGHT

Except where otherwise noted, I<Class::ErrorHandler> is Copyright 2004
Benjamin Trott, cpan@stupidfool.org. All rights reserved.

=cut
